### DISCUSSION TITLE
New api tokens not working?

### DISCUSSION THREAD

**Discussion by tofuSCHNITZEL** [kimai/kimai#4750] [Questions & Answers] [answered]
I tried to switch my scripts to the new bearer tokens and I followed the API examples but I always get:
```
  Client error: `POST http://kimai.test/api/customers` resulted in a `401 Unauthorized` response:
  {"code":401,"message":"Full authentication is required to access this resource."}
```
I even copied the code from the API example 1:1 and just inserted my token but I still get the "Full authentication required" error

What am I missing here? the "KIMAI_API_TOKEN" should just be the token displayed once when creating it, right?

**Comment by kevinpapst** [comment]
Please share what not works. This does:

<img width="773" alt="Bildschirmfoto 2024-04-06 um 23 36 19" src="https://github.com/kimai/kimai/assets/533162/fd429cc7-1644-4165-a93c-f70c5ecc4b97">

```
curl -X "POST" "https://127.0.0.1:8016/api/projects" \
     -H 'Content-Type: application/json' \
     -H 'Authorization: Bearer 2dc9b3d77b7641c955d911c8c' \
     -d $'{
  "start": "2023-02-07T11:02:25",
  "customer": 6,
  "comment": "sdfsdf",
  "name": "API project",
  "visible": true
}'
```

**Reply by tofuSCHNITZEL** [reply]
no matter what api endpoint I query I always get the 401 error:

```
curl -X "GET" "http://kimai.test/api/ping" -H 'Content-Type: application/json' -H 'Authorization: Bearer 56d7ed0a06ed9dbb5a75f3d8e'

{"code":401,"message":"Full authentication is required to access this resource."}
```
Insomnia:
![image](https://github.com/kimai/kimai/assets/5708088/a75090d6-981e-46cb-abc0-24d4829e0982)
or:
![image](https://github.com/kimai/kimai/assets/5708088/ae0ff373-cbb7-4ea7-bfda-b8c60467d5c0)

token exists in database:
![image](https://github.com/kimai/kimai/assets/5708088/6085aad2-7fe3-495d-aca2-b41dd3c583e0)

**Reply by kevinpapst** [reply]
I am not able to reproduce. It works for me:
```
curl "https://demo.kimai.org/api/users/me" -H 'Authorization: Bearer 1c41872cabfe945915bef6c0d'
```
Can you try with the demo?

```bash
$ curl -X "GET" "https://demo.kimai.org/api/ping" -H 'Content-Type: application/json' -H 'Authorization: Bearer 1c41872cabfe945915bef6c0d'

{"message":"pong"}
```

**Reply by tofuSCHNITZEL** [reply]
yes demo works for me also with a new bearer token I created...
could it have something to do with my local test install only running under http and not https?

**Reply by kevinpapst** [reply]
maybe, all my system run on https only, i never tried with http. 

I'd say I try to reproduce, but https is insecure by design and I guess it might be better if you try to upgrade your test system?!

**Reply by tofuSCHNITZEL** [reply]
I have tried it with a https install now. there I also get an 401 error but with a simpler error message:
`  {"code":401,"message":"Unauthorized"}`

**Reply by h3nr1-g** [reply]
I'm experiencing the same problem with the docker image of the latest release. 

Token values are stored correctly in the DB but every try with curl, python requests, ... ends in the same result as described by tofuSCHNITZEL.

**Reply by kevinpapst** [reply]
And does the "old API password" work when you execute the same call, but just swap the headers?

**Reply by tofuSCHNITZEL** [reply]
yes the old way still works without a problem

**Reply by kevinpapst** [reply]
do you see the problem with `http` or also with `https`?

**Reply by h3nr1-g** [reply]
I use a nginx as a reverse proxy to provide to the application via http and https. It makes no difference.

**Reply by tofuSCHNITZEL** [reply]
yes I can also confirm neither http nor https changes anything I have tested it with instances hosted by apache and instances hosted by an IIS

**Reply by h3nr1-g** [reply]
Hey, 
I took a deeper look into the logs of the docker container and found following exception message(s):

`kimai-kimai-1    | [2024-04-09T20:52:07.505586+00:00] request.INFO: Matched route "get_users". {"route":"get_users","route_parameters":{"_route":"get_users","_controller":"App\\API\\UserController::cgetAction"},"request_uri":"http://127.0.0.1:8001/api/users","method":"GET"} {"channel":"request"}

kimai-kimai-1    | [2024-04-09T20:52:07.512066+00:00] security.NOTICE: No Authentication entry point configured, returning a 401 HTTP response. Configure "entry_point" on the firewall "api" if you want to modify the response. [] {"channel":"security"}

kimai-kimai-1    | [2024-04-09T20:52:07.512357+00:00] request.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\HttpException: "Full authentication is required to access this resource." at ExceptionListener.php line 232 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\HttpException(code: 0): Full authentication is required to access this resource. at /opt/kimai/vendor/symfony/security-http/Firewall/ExceptionListener.php:232)\n[previous exception] [object] (Symfony\\Component\\Security\\Core\\Exception\\InsufficientAuthenticationException(code: 0): Full authentication is required to access this resource. at /opt/kimai/vendor/symfony/security-http/Firewall/ExceptionListener.php:143)\n[previous exception] [object] (Symfony\\Component\\Security\\Core\\Exception\\AccessDeniedException(code: 403): Access Denied. at /opt/kimai/vendor/symfony/security-http/Firewall/AccessListener.php:87)"} {"channel":"request"}`

**Reply by tofuSCHNITZEL** [reply]
okay I think I have an idea what might be going on...
it seems it only works if I'm also logged in.

if I use http://kimai.test/api/doc and enter my bearer token in the top right "authorize" and then try e.g. the ping it works.
If I run the curl statement it provides I get "{"code":401,"message":"Full authentication is required to access this resource."}"
BUT if I then edit the curl command to include my session cookie from the request the api test site sent (-H 'Cookie: PHPSESSID=i0lbtg467p92rg636fmnmki9bm; ' ) - it also works with the curl statement..

**Reply by kevinpapst** [reply]
> if I use http://kimai.test/api/doc and enter my bearer token in the top right "authorize" and then try e.g. the ping it works.

Okay, so you see something like 
```
curl -X 'GET' \
  'http://127.0.0.1:8001/api/ping' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer xyzxyzyxzyxzyzxz'
```

And that works via API docs. But this curl then fails when used via command line. 

> if I then edit the curl command to include my session cookie

That's to be expected, it re-uses your existing session.

Some research on that error message. It is thrown in [Symfony if a user if not "fullyFledged"](https://github.com/symfony/symfony/blob/7.1/src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php#L143). I am not sure yet what that means. And why it happens for your test-case but not mine.

I tested with SAML accounts and local accounts. I tested with and without activated two-factor. I truncated the kimai2_sessions table to make sure that no session is re-used. It just works for me. And in all systems that I own. 
Can you think of any other cause for that issue on your end?

**Reply by kevinpapst** [reply]
Okay, I found out that these exceptions are thrown if the Authorization headers are missing completely or if e.g. `bearer` is written in the wrong case.

So I can safely reproduce the issue if I simply skip the Authorization header or write `Authorization: bearer 2dc9b3d77b7641c955d911c8c` (instead of `Authorization: Bearer 2dc9b3d77b7641c955d911c8c`).

Which lets me believe, that there is something fishy with your installations.

@tofuSCHNITZEL you said it is working in the demo, but not on your system. Is it docker? a reverse proxy? Maybe the proxy is not forwarding the headers (properly)?

**Reply by tofuSCHNITZEL** [reply]
my local test install is a Laragon (apache) on windows.
![image](https://github.com/kimai/kimai/assets/5708088/dfc46b15-22eb-49e1-b210-faac84db92de)

the other installation with the issue is on linux with plesk hosting so its an apache backend (php-fpm with apache) with an nginx in front 

and the third install is on windows IIS.

I consider the laragon installation to be the most "normal" one... there is nothing in front of it that should tamper with the headers... do you have a tip where to add breakpoints to see if the header was changed somewhere?

is it feasable to make the whole thing case insensitive?

**Reply by kevinpapst** [reply]
> is it feasable to make the whole thing case insensitive?

No, we are talking about a global HTTP standard here.

Can you edit `public/index.php` and add this in the first line:
```
var_dump($_SERVER['HTTP_AUTHORIZATION']);exit;
```
<img width="701" alt="Bildschirmfoto 2024-04-12 um 14 24 22" src="https://github.com/kimai/kimai/assets/533162/2404e9ba-afe7-4829-a8c8-9724ef4964a5">

The result of `curl "https://127.0.0.1:8016/api/users/me" -H 'Authorization: Bearer 1c41872cabfe945915bef6c0d' --insecure` should be:
````
string(32) "Bearer 1c41872cabfe945915bef6c0d"
````
<img width="1196" alt="Bildschirmfoto 2024-04-12 um 14 26 12" src="https://github.com/kimai/kimai/assets/533162/f02d67e3-4c93-4e3f-b763-dc52db9fc83f">

**Reply by tofuSCHNITZEL** [reply]
okay interesting... I only get Undefined array key "HTTP_AUTHORIZATION"

then I added:
`SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1`
(ref: https://www.wp-pay.org/http-authorization-header-missing/)
to the .htaccess file and now everything is working.

so it seems something in all of my different installations is stripping the authorization header away

-
I also added this to the htaccess on my plesk/apache/nginx server and now it works there as well...
And for IIS it seems to just work without any additional tricks.

**Reply by kevinpapst** [reply]
Weird that Apache strips it, never heard of that before.

I guess I should add that to the documentation pages, as very likely more people will show up with that issue.

So at the end we landed at: Hoorray, fixed 😁

**Reply by tofuSCHNITZEL** [reply]
And should we add "SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1" to the official htaccess?
seems to me that it would do no harm if not needed?

**Reply by tofuSCHNITZEL** [reply, marked as answer]
Adding
`SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1`
to .htaccess seems to solve the problem.

**Comment by kevinpapst** [comment]
I created a bug here to follow: https://github.com/kimai/kimai/issues/4774

The error now shows up in my logs as well. 
I still have no idea how to reproduce the issue though.
What users did you use for testing? System-Admins? Regular users? 
Can you share your docker-compose so I can test it with your setup?

**Comment by tofuSCHNITZEL** [comment]
Adding
`SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1`
to .htaccess seems to solve the problem.

**Reply by kevinpapst** [reply]
Hm, good idea. I just checked the official Symfony recipe:
https://github.com/symfony/recipes-contrib/blob/main/symfony/apache-pack/1.0/public/.htaccess#L38

Seems I didn't merge these files for a long time.
I'd like to stick as close to the "Standard".
Could you give that a try and see if it works as well?

**Reply by tofuSCHNITZEL** [reply]
yes!
I replaced the existing htaccess with this one and it also works! 
so you only need to merge this upstream htaccess and we are good to go ;)

**Reply by kevinpapst** [reply]
Already added to the next release:
https://github.com/kimai/kimai/pull/4749/commits/f8ef52578f6864baa6c12aabb10ba834d39d5a5d

Sometimes it is really frustrating to find and fix such environment specific issues.
I wouldn't have found the cause without you.

Big thanks for your support 👍

**Reply by tofuSCHNITZEL** [reply]
I totally understand... Thank you for taking the time and the patience to get to the bottom of this!

**Comment by prathers** [comment]
I can see that this on os closed, but I'll try anyway:

I'm running kimai under NGINX and as far as I know, .htraccess will not be taken into account under this environment. I face the same problem:

`curl --location 'https://[domain]/api/ping' -H 'Authorization: bearer [token]'
{"code":401,"message":"Unauthorized"}%` 

Is there a way to solve it on a NGINX server..?

**Reply by kevinpapst** [reply]
- Wrong: `Authorization: bearer [token]`
- Correct: `Authorization: Bearer [token]`
